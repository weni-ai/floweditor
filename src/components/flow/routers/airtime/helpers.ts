import {
  AirtimeRouterFormState,
  AirtimeTransferEntry,
} from 'components/flow/routers/airtime/AirtimeRouterForm';
import { createWebhookBasedNode } from 'components/flow/routers/helpers';
import { Types } from 'config/interfaces';
import { TransferAirtime } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings } from 'store/nodeEditor';
import { createUUID } from 'utils';

export const nodeToState = (
  settings: NodeEditorSettings,
): AirtimeRouterFormState => {
  const originalAction = getOriginalAction(settings);
  let resultName = { value: 'Result' };
  let valid = false;

  const amounts: AirtimeTransferEntry[] = [];
  if (originalAction && originalAction.type === Types.transfer_airtime) {
    Object.keys(originalAction.amounts).forEach((key: string) => {
      amounts.push({
        value: { code: key, amount: '' + originalAction.amounts[key] },
      });
    });
    resultName = { value: originalAction.result_name };
    valid = true;
  }

  return {
    valid,
    amounts,
    resultName,
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: AirtimeRouterFormState,
): RenderNode => {
  let uuid = createUUID();
  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }

  const amounts = {};
  state.amounts.forEach((entry: AirtimeTransferEntry) => {
    if (entry.value.amount.trim().length > 0) {
      (amounts as any)[entry.value.code] = Number(entry.value.amount);
    }
  });

  const newAction: TransferAirtime = {
    uuid,
    type: Types.transfer_airtime,
    amounts,
    result_name: state.resultName.value,
  };

  return createWebhookBasedNode(newAction, settings.originalNode, true);
};

export const getOriginalAction = (
  settings: NodeEditorSettings,
): TransferAirtime => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 &&
      settings.originalNode.node.actions[0]);

  if (action.type === Types.transfer_airtime) {
    return action as TransferAirtime;
  }
};
