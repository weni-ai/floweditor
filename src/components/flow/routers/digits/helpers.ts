import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import {
  createCaseProps,
  createRenderNode,
  hasCases,
  resolveRoutes,
} from 'components/flow/routers/helpers';
import { Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import {
  HintTypes,
  Router,
  RouterTypes,
  SwitchRouter,
  WaitTypes,
} from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';

import { DigitsRouterFormState } from './DigitsRouterForm';
import { DIGITS_OPERAND } from 'components/nodeeditor/constants';

export const nodeToState = (
  settings: NodeEditorSettings,
): DigitsRouterFormState => {
  let initialCases: CaseProps[] = [];

  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: '' };

  if (getType(settings.originalNode) === Types.wait_for_digits) {
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      if (hasCases(settings.originalNode.node)) {
        initialCases = createCaseProps(router.cases, settings.originalNode);
      }

      resultName = { value: router.result_name || '' };
    }
  }

  return {
    cases: initialCases,
    resultName,
    valid: true,
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: DigitsRouterFormState,
): RenderNode => {
  const {
    cases,
    exits,
    categories,
    defaultCategory: defaultExit,
    caseConfig,
  } = resolveRoutes(state.cases, false, settings.originalNode.node);

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    default_category_uuid: defaultExit,
    categories,
    cases,
    operand: DIGITS_OPERAND,
    wait: { type: WaitTypes.msg, hint: { type: HintTypes.digits } },
    ...optionalRouter,
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.wait_for_response,
    [],

    { cases: caseConfig },
  );

  return newRenderNode;
};
