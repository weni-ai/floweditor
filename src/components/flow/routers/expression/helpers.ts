import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import { ExpressionRouterFormState } from 'components/flow/routers/expression/ExpressionRouterForm';
import {
  createCaseProps,
  createRenderNode,
  hasCases,
  resolveRoutes,
} from 'components/flow/routers/helpers';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { Router, RouterTypes, SwitchRouter } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';

export const nodeToState = (
  settings: NodeEditorSettings,
): ExpressionRouterFormState => {
  let initialCases: CaseProps[] = [];

  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: '' };

  let operand = DEFAULT_OPERAND;

  if (
    settings.originalNode &&
    getType(settings.originalNode) === Types.split_by_expression
  ) {
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      if (hasCases(settings.originalNode.node)) {
        initialCases = createCaseProps(router.cases, settings.originalNode);
      }

      resultName = { value: router.result_name || '' };
    }

    operand = router.operand;
  }

  return {
    cases: initialCases,
    resultName,
    operand: { value: operand },
    valid: true,
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: ExpressionRouterFormState,
): RenderNode => {
  const {
    cases,
    exits,
    defaultCategory: defaultExit,
    caseConfig,
    categories,
  } = resolveRoutes(state.cases, false, settings.originalNode.node);

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    default_category_uuid: defaultExit,
    categories,
    cases,
    operand: state.operand.value,
    ...optionalRouter,
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.split_by_expression,
    [],
    { cases: caseConfig },
  );

  return newRenderNode;
};
