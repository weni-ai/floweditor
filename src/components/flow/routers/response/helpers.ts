import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import {
  createCaseProps,
  createRenderNode,
  hasCases,
  resolveRoutes,
} from 'components/flow/routers/helpers';
import { ResponseRouterFormState } from 'components/flow/routers/response/ResponseRouterForm';
import { DEFAULT_OPERAND } from 'components/nodeeditor/constants';
import { Type, Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { Router, RouterTypes, SwitchRouter, Wait, WaitTypes } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, StringEntry } from 'store/nodeEditor';

export const nodeToState = (
  settings: NodeEditorSettings,
): ResponseRouterFormState => {
  let initialCases: CaseProps[] = [];

  // TODO: work out an incremental result name
  let resultName: StringEntry = { value: 'Result' };
  let timeout = 0;

  if (
    (settings.originalNode &&
      getType(settings.originalNode) === Types.wait_for_response) ||
    getType(settings.originalNode) === Types.smart_wait_for_response
  ) {
    const router = settings.originalNode.node.router as SwitchRouter;
    if (router) {
      if (hasCases(settings.originalNode.node)) {
        initialCases = createCaseProps(router.cases, settings.originalNode);
      }

      resultName = { value: router.result_name || '' };
    }

    if (
      settings.originalNode.node.router.wait &&
      settings.originalNode.node.router.wait.timeout
    ) {
      timeout = settings.originalNode.node.router.wait.timeout.seconds || 0;
    }
  }

  return {
    cases: initialCases,
    resultName,
    timeout,
    valid: true,
  };
};

export const stateToNode = (
  settings: NodeEditorSettings,
  typeConfig: Type,
  state: ResponseRouterFormState,
): RenderNode => {
  const {
    cases,
    exits,
    defaultCategory,
    timeoutCategory,
    caseConfig,
    categories,
  } = resolveRoutes(state.cases, state.timeout > 0, settings.originalNode.node);

  const optionalRouter: Pick<Router, 'result_name'> = {};
  if (state.resultName.value) {
    optionalRouter.result_name = state.resultName.value;
  }

  const wait = { type: WaitTypes.msg } as Wait;
  if (state.timeout > 0) {
    wait.timeout = {
      seconds: state.timeout,
      category_uuid: timeoutCategory,
    };
  }

  const router: SwitchRouter = {
    type:
      typeConfig.type === Types.smart_wait_for_response
        ? RouterTypes.smart
        : RouterTypes.switch,
    default_category_uuid: defaultCategory,
    cases,
    categories,
    operand: DEFAULT_OPERAND,
    wait,
    ...optionalRouter,
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    typeConfig.type,
    [],
    { cases: caseConfig },
  );

  return newRenderNode;
};
