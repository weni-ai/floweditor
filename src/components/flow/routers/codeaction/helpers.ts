import { createRenderNode } from 'components/flow/routers/helpers';
import {
  Category,
  Exit,
  RouterTypes,
  SwitchRouter,
  WebhookExitNames,
} from 'flowTypes';
import { Types } from 'config/interfaces';
import { RunCodeAction } from 'flowTypes';
import { RenderNode } from 'store/flowContext';
import { NodeEditorSettings, FormEntry } from 'store/nodeEditor';
import { createUUID, snakify } from 'utils';
import { CodeActionRouterFormState } from 'components/flow/routers/codeaction/CodeActionRouterForm';
import { StringEntry } from '../../../../store/nodeEditor';

export const getOriginalAction = (
  settings: NodeEditorSettings,
): RunCodeAction => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 &&
      settings.originalNode.node.actions[0]);

  if (action.type === Types.run_code_action) {
    return action as RunCodeAction;
  }
};

export const nodeToState = (
  settings: NodeEditorSettings,
  initialCodeAction: any,
): CodeActionRouterFormState => {
  const originalAction = getOriginalAction(settings);
  let resultName: StringEntry = { value: 'Result' };
  let codeAction: FormEntry = {
    value: { id: 0, name: 'Select a Code Action' },
  };

  if (originalAction && originalAction.type === Types.run_code_action) {
    codeAction = { value: originalAction.codeaction };
    resultName = { value: originalAction.result_name };
  } else {
    if (initialCodeAction) {
      codeAction = initialCodeAction;
    }
  }

  const state: CodeActionRouterFormState = {
    resultName,
    codeAction,
    valid: true,
  };
  return state;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: CodeActionRouterFormState,
): RenderNode => {
  let uuid = createUUID();
  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }
  let categories: Category[] = [];
  let exits: Exit[] = [];

  const newAction: RunCodeAction = {
    uuid,
    type: Types.run_code_action,
    codeaction: {
      id: state.codeAction.value.id,
      name: state.codeAction.value.name,
    },
    result_name: state.resultName.value,
  };

  const originalNode = settings.originalNode;
  if (
    originalNode &&
    originalNode.node.actions.length === 1 &&
    originalNode.node.actions[0].type === newAction.type
  ) {
    exits = originalNode.node.exits;
    categories = originalNode.node.router.categories;
  } else {
    exits = [
      {
        uuid: createUUID(),
        destination_uuid: null,
      },
      {
        uuid: createUUID(),
        destination_uuid: null,
      },
    ];

    categories = [
      {
        uuid: createUUID(),
        name: WebhookExitNames.Success,
        exit_uuid: exits[0].uuid,
      },
      {
        uuid: createUUID(),
        name: WebhookExitNames.Failure,
        exit_uuid: exits[1].uuid,
      },
    ];
  }

  let operand = '@results.' + snakify(newAction.result_name);

  const router: SwitchRouter = {
    type: RouterTypes.switch,
    default_category_uuid: categories[0].uuid,
    cases: [],
    categories,
    operand,
  };

  const newRenderNode = createRenderNode(
    settings.originalNode.node.uuid,
    router,
    exits,
    Types.run_code_action,
    [newAction],
  );

  return newRenderNode;
};
