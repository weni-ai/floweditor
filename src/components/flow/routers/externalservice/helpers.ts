import { createWebhookBasedNode } from 'components/flow/routers/helpers';
import { ServiceCall, Types } from 'config/interfaces';
import { getType } from 'config/typeConfigs';
import { CallExternalService } from 'flowTypes';
import { AssetMap, RenderNode } from 'store/flowContext';
import { NodeEditorSettings, FormEntry } from 'store/nodeEditor';
import { createUUID } from 'utils';
import { ExternalServiceRouterFormState } from 'components/flow/routers/externalservice/ExternalServiceRouterForm';

export const getOriginalAction = (
  settings: NodeEditorSettings,
): CallExternalService => {
  const action =
    settings.originalAction ||
    (settings.originalNode.node.actions.length > 0 &&
      settings.originalNode.node.actions[0]);

  if (action.type === Types.call_external_service) {
    return action as CallExternalService;
  }
};

export const nodeToState = (
  settings: NodeEditorSettings,
  existingServices: AssetMap,
  initialExternalService: any,
): ExternalServiceRouterFormState => {
  let resultName = { value: 'Result' };

  let externalService: FormEntry = null;
  let initialCall: FormEntry = null;
  let initialParams: any[] = [];

  if (
    settings.originalNode &&
    getType(settings.originalNode) === Types.split_by_external_service
  ) {
    const action = getOriginalAction(settings) as CallExternalService;
    const service = existingServices[action.external_service.uuid];

    if (service) {
      externalService = {
        value: {
          ...service.content,
          name: service.name,
          external_service_type: service.type,
          uuid: service.id,
        },
      };
    } else {
      externalService = { value: action.external_service };
    }

    const existingCallParams =
      externalService.value.actions &&
      externalService.value.actions.find((a: ServiceCall) => {
        return a.name === action.call.name;
      }).params;

    initialCall = {
      value: existingCallParams
        ? { ...action.call, params: existingCallParams }
        : action.call,
    };
    initialParams = action.params;
    resultName = { value: action.result_name };
  } else {
    externalService = initialExternalService
      ? {
          value: {
            uuid: initialExternalService.id,
            name: initialExternalService.name,
            type: initialExternalService.type,
            actions: initialExternalService.content.actions,
          },
        }
      : { value: null };
  }

  const state: ExternalServiceRouterFormState = {
    externalService,
    call: initialCall,
    resultName,
    params: { value: initialParams },
    valid: true,
  };

  return state;
};

export const stateToNode = (
  settings: NodeEditorSettings,
  state: ExternalServiceRouterFormState,
): RenderNode => {
  let uuid = createUUID();
  const originalAction = getOriginalAction(settings);
  if (originalAction) {
    uuid = originalAction.uuid;
  }

  const validParams = state.params.value.filter(
    (param: any) =>
      param.valid === true &&
      param.data.value &&
      param.data.validationFailures.length === 0,
  );

  const newAction: CallExternalService = {
    uuid,
    type: Types.call_external_service,
    external_service: {
      uuid: state.externalService.value.uuid,
      name: state.externalService.value.name,
      external_service_type:
        state.externalService.value.external_service_type ||
        state.externalService.value.type,
      actions: state.externalService.value.actions,
    },
    call: state.call.value,
    params: validParams,
    result_name: state.resultName.value,
  };

  return createWebhookBasedNode(newAction, settings.originalNode, true);
};
