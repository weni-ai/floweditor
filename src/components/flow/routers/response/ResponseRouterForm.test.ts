import { RouterFormProps } from 'components/flow/props';
import { CaseProps } from 'components/flow/routers/caselist/CaseList';
import ResponseRouterForm from 'components/flow/routers/response/ResponseRouterForm';
import { Operators } from 'config/interfaces';
import { Types } from 'config/interfaces';
import { composeComponentTestUtils, mock } from 'testUtils';
import { getRouterFormProps, createMatchRouter } from 'testUtils/assetCreators';
import * as utils from 'utils';
import { createUUID } from 'utils';
import { getSmartOrSwitchRouter } from 'components/flow/routers/helpers';
import { shallowToJson } from 'enzyme-to-json';

const routerNode = createMatchRouter(['Red']);

const { setup } = composeComponentTestUtils<RouterFormProps>(
  ResponseRouterForm,
  getRouterFormProps(routerNode),
);

describe(ResponseRouterForm.name, () => {
  beforeEach(() => {
    mock(utils, 'createUUID', utils.seededUUIDs());
  });

  it('should render', () => {
    const { wrapper } = setup(true);
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('initializes case config', () => {
    const dateCase = createUUID();

    const dateNode = createMatchRouter(
      ['Red'],
      null,
      Operators.has_any_word,
      null,
      true,
    );
    const router = getSmartOrSwitchRouter(dateNode.node);
    router.cases.push({
      uuid: dateCase,
      type: Operators.has_date_eq,
      arguments: ['@(datetime_add(today(), 5, "D"))'],
      category_uuid: router.categories[0].uuid,
    });

    dateNode.ui = {
      position: { left: 0, top: 0 },
      type: Types.wait_for_response,
      config: { cases: { [dateCase]: { arguments: ['5'] } } },
    };

    const { wrapper } = setup(true, {
      nodeSettings: {
        $set: {
          originalNode: dateNode,
        },
      },
    });

    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  describe('updates', () => {
    it('should save changes', () => {
      const { instance, props } = setup(true, {
        $merge: { onClose: vi.fn(), updateRouter: vi.fn() },
      });

      instance.handleUpdateTimeout(180);
      instance.handleUpdateResultName('Favorite Color');
      instance.handleCasesUpdated([
        {
          uuid: createUUID(),
          kase: { type: Operators.has_any_word, arguments: ['red'] },
          categoryName: 'Red',
          valid: true,
        },
        {
          uuid: createUUID(),
          kase: { type: Operators.has_any_word, arguments: ['maroon'] },
          categoryName: 'Red',
          valid: true,
        },
        {
          uuid: createUUID(),
          kase: { type: Operators.has_any_word, arguments: ['green'] },
          categoryName: 'Green',
          valid: true,
        },
      ] as CaseProps[]);

      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).toHaveBeenCalled();
      expect(props.updateRouter).toMatchSnapshot();
    });

    it('should save save config for relative dates', () => {
      const { instance, props } = setup(true, {
        $merge: { onClose: vi.fn(), updateRouter: vi.fn() },
      });

      instance.handleCasesUpdated([
        {
          uuid: createUUID(),
          kase: { type: Operators.has_date_gt, arguments: ['5'] },
          categoryName: 'In the Zone',
          valid: true,
        },
      ] as CaseProps[]);

      expect(instance.state).toMatchSnapshot();

      instance.handleSave();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).toHaveBeenCalled();
      expect(props.updateRouter).toMatchSnapshot();
    });

    it('should cancel', () => {
      const { instance, props } = setup(true, {
        $merge: { onClose: vi.fn(), updateRouter: vi.fn() },
      });
      instance.handleUpdateTimeout(180);
      instance.getButtons().secondary.onClick();
      expect(props.onClose).toHaveBeenCalled();
      expect(props.updateRouter).not.toHaveBeenCalled();
    });
  });
});
