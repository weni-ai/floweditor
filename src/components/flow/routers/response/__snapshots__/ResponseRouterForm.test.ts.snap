// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResponseRouterForm > initializes case config 1`] = `
<Dialog
  buttons={
    {
      "primary": {
        "name": "Confirm",
        "onClick": [Function],
      },
      "secondary": {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  gutter={
    <TimeoutControl
      onChanged={[Function]}
      timeout={60}
    />
  }
  headerClass="wait_for_response"
  title="Wait for response"
>
  <TypeList
    __className=""
    initialType={
      {
        "aliases": [
          "switch",
        ],
        "description": "Wait for the contact to respond",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": [
          "categories",
          "cases",
        ],
        "name": "Wait for response",
        "type": "wait_for_response",
        "visibility": [
          "messaging",
          "messaging_offline",
        ],
      }
    }
    nodeSettings={
      {
        "originalNode": {
          "inboundConnections": {},
          "node": {
            "actions": [],
            "exits": [
              {
                "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
              },
              {
                "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
              },
              {
                "uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
              },
            ],
            "router": {
              "cases": [
                {
                  "arguments": [
                    "red",
                  ],
                  "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                  "type": "has_any_word",
                  "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                },
                {
                  "arguments": [
                    "@(datetime_add(today(), 5, "D"))",
                  ],
                  "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                  "type": "has_date_eq",
                  "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
                },
              ],
              "categories": [
                {
                  "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                  "name": "Red",
                  "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                },
                {
                  "exit_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
                  "name": "Other",
                  "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
                },
                {
                  "exit_uuid": "2dc85899-0577-4f1b-a620-f12094e34b5e",
                  "name": "No Response",
                  "uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
                },
              ],
              "default_category_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
              "operand": null,
              "result_name": null,
              "type": "switch",
              "wait": {
                "timeout": {
                  "category_uuid": "42ebf1d0-a43e-4ba6-9577-c863b20c88c4",
                  "seconds": 60,
                },
                "type": "msg",
              },
            },
            "uuid": "70ac6ea6-803a-4c33-81c7-dc26803c313f",
          },
          "ui": {
            "config": {
              "cases": {
                "b1f332f3-bdd3-4891-aec5-1843a712dbf1": {
                  "arguments": [
                    "5",
                  ],
                },
              },
            },
            "position": {
              "left": 0,
              "top": 0,
            },
            "type": "wait_for_response",
          },
        },
      }
    }
    onChange={[MockFunction spy]}
  />
  <div>
    If the message response
  </div>
  <CaseList
    cases={
      [
        {
          "categoryName": "Red",
          "kase": {
            "arguments": [
              "red",
            ],
            "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "type": "has_any_word",
            "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
          },
          "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
          "valid": true,
        },
        {
          "categoryName": "Red",
          "kase": {
            "arguments": [
              "5",
            ],
            "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            "type": "has_date_eq",
            "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
          },
          "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
          "valid": true,
        },
      ]
    }
    data-spec="cases"
    onCasesUpdated={[Function]}
    type="default"
  />
  <OptionalTextInput
    maxLength={64}
    name="Save as result"
    onChange={[Function]}
    toggleText="Save as"
    value={
      {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`ResponseRouterForm > should render 1`] = `
<Dialog
  buttons={
    {
      "primary": {
        "name": "Confirm",
        "onClick": [Function],
      },
      "secondary": {
        "name": "Cancel",
        "onClick": [Function],
      },
    }
  }
  gutter={
    <TimeoutControl
      onChanged={[Function]}
      timeout={0}
    />
  }
  headerClass="wait_for_response"
  title="Wait for response"
>
  <TypeList
    __className=""
    initialType={
      {
        "aliases": [
          "switch",
        ],
        "description": "Wait for the contact to respond",
        "form": [Function],
        "localization": [Function],
        "localizeableKeys": [
          "categories",
          "cases",
        ],
        "name": "Wait for response",
        "type": "wait_for_response",
        "visibility": [
          "messaging",
          "messaging_offline",
        ],
      }
    }
    nodeSettings={
      {
        "originalAction": null,
        "originalNode": {
          "inboundConnections": {},
          "node": {
            "actions": [],
            "exits": [
              {
                "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
              },
              {
                "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
              },
            ],
            "router": {
              "cases": [
                {
                  "arguments": [
                    "red",
                  ],
                  "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                  "type": "has_any_word",
                  "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
                },
              ],
              "categories": [
                {
                  "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                  "name": "Red",
                  "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                },
                {
                  "exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                  "name": "Other",
                  "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
                },
              ],
              "default_category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
              "operand": "@input.text",
              "result_name": "",
              "type": "switch",
              "wait": {
                "type": "msg",
              },
            },
            "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
          },
          "ui": {
            "position": {
              "left": 0,
              "top": 0,
            },
            "type": "wait_for_response",
          },
        },
      }
    }
    onChange={[MockFunction spy]}
  />
  <div>
    If the message response
  </div>
  <CaseList
    cases={
      [
        {
          "categoryName": "Red",
          "kase": {
            "arguments": [
              "red",
            ],
            "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            "type": "has_any_word",
            "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          },
          "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
          "valid": true,
        },
      ]
    }
    data-spec="cases"
    onCasesUpdated={[Function]}
    type="default"
  />
  <OptionalTextInput
    maxLength={64}
    name="Save as result"
    onChange={[Function]}
    toggleText="Save as"
    value={
      {
        "value": "",
      }
    }
  />
</Dialog>
`;

exports[`ResponseRouterForm > updates > should save changes 1`] = `
{
  "cases": [
    {
      "categoryName": "Red",
      "kase": {
        "arguments": [
          "red",
        ],
        "type": "has_any_word",
      },
      "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
      "valid": true,
    },
    {
      "categoryName": "Red",
      "kase": {
        "arguments": [
          "maroon",
        ],
        "type": "has_any_word",
      },
      "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
      "valid": true,
    },
    {
      "categoryName": "Green",
      "kase": {
        "arguments": [
          "green",
        ],
        "type": "has_any_word",
      },
      "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
      "valid": true,
    },
  ],
  "resultName": {
    "validationFailures": [],
    "value": "Favorite Color",
  },
  "timeout": 180,
  "valid": true,
}
`;

exports[`ResponseRouterForm > updates > should save changes 2`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "inboundConnections": {},
        "node": {
          "actions": [],
          "exits": [
            {
              "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
            },
            {
              "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
            },
            {
              "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            },
            {
              "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            },
          ],
          "router": {
            "cases": [
              {
                "arguments": [
                  "red",
                ],
                "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                "type": "has_any_word",
              },
              {
                "arguments": [
                  "maroon",
                ],
                "category_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                "type": "has_any_word",
              },
              {
                "arguments": [
                  "green",
                ],
                "category_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                "type": "has_any_word",
              },
            ],
            "categories": [
              {
                "exit_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                "name": "Red",
                "uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
              },
              {
                "exit_uuid": "763e4844-3e1b-407a-a1b5-5fdfcd308b41",
                "name": "Green",
                "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
              },
              {
                "exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                "name": "Other",
                "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
              },
              {
                "exit_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
                "name": "No Response",
                "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
              },
            ],
            "default_category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "operand": "@input.text",
            "result_name": "Favorite Color",
            "type": "switch",
            "wait": {
              "timeout": {
                "category_uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
                "seconds": 180,
              },
              "type": "msg",
            },
          },
          "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
        },
        "ui": {
          "config": {
            "cases": {},
          },
          "position": null,
          "type": "wait_for_response",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ResponseRouterForm > updates > should save save config for relative dates 1`] = `
{
  "cases": [
    {
      "categoryName": "In the Zone",
      "kase": {
        "arguments": [
          "5",
        ],
        "type": "has_date_gt",
      },
      "uuid": "b1f332f3-bdd3-4891-aec5-1843a712dbf1",
      "valid": true,
    },
  ],
  "resultName": {
    "value": "",
  },
  "timeout": 0,
  "valid": true,
}
`;

exports[`ResponseRouterForm > updates > should save save config for relative dates 2`] = `
[MockFunction spy] {
  "calls": [
    [
      {
        "inboundConnections": {},
        "node": {
          "actions": [],
          "exits": [
            {
              "uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
            },
            {
              "uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
            },
          ],
          "router": {
            "cases": [
              {
                "arguments": [
                  "@(datetime_add(today(), 5, "D"))",
                ],
                "category_uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
                "type": "has_date_gt",
              },
            ],
            "categories": [
              {
                "exit_uuid": "061fc171-8b79-4636-b892-bd0ea5aa9b42",
                "name": "In the Zone",
                "uuid": "1dce2b34-9aab-4e20-81c4-3f0408dcb671",
              },
              {
                "exit_uuid": "477ac8b4-25e2-483a-8686-2d1332c4da1c",
                "name": "Other",
                "uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
              },
            ],
            "default_category_uuid": "1e47a1e1-3c67-4df5-adf1-da542c789adb",
            "operand": "@input.text",
            "type": "switch",
            "wait": {
              "type": "msg",
            },
          },
          "uuid": "132de855-4042-4dc1-a18f-cc2e6a8f790a",
        },
        "ui": {
          "config": {
            "cases": {
              "b1f332f3-bdd3-4891-aec5-1843a712dbf1": {
                "arguments": [
                  "5",
                ],
              },
            },
          },
          "position": null,
          "type": "wait_for_response",
        },
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
